### ISSUE TITLE
Audio Mixers - Live changes

### ISSUE THREAD

**Issue by s9menine** [godotengine/godot-proposals#1186]
**Describe the project you are working on:**
As a sound designer, all my projects feature extensive use of the audio mixer, buss routing, and effects.

**Describe the problem or limitation you are having in your project:**
Fine-tuning audio balance between mixer busses and effect parameters while the game is running is currently not possible. Given that this is an iterative process, having to reload scenes to test every single audio mixer change turning it into trial-and-error is painfully ineffiecient.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Sync audio mixer adjustments while the game is running, similar to how changing scene objects or script variables are synced.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Unity has a toggle option on their audio mixer:
![Unity_SDuJvrsSy1](https://user-images.githubusercontent.com/40716090/87181004-296d1b80-c314-11ea-8d6d-bf09d946d864.png)
Activating it syncs changes to the audio mixer to the running scene, immediately audible; the audio mixer's meter graphics start responding to the game's audio levels as well.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Not a few lines, but a technically possible workaround is creating a UI in-game that uses AudioServer methods to modify audio mixers and their effects, basically recreating the entire Godot project's audio mixer into the game. After achieving an acceptable result, write down values and then adjust the actual audio mixer accordingly. 

**Is there a reason why this should be core and not an add-on in the asset library?:**
The audio mixer seems to be a core Godot feature; I can't find add-ons or plugins that can modify the audio mixer.

**Comment by Calinou** [comment]
cc @Faless - Do you think this should warrant a new option in the Debug menu, or should we just reuse the value of **Sync Scene Changes**?

**Comment by s9menine** [comment]
(Don't know if the question was aimed at me, but:)

Personally I would always want to sync both scene changes and audio bus editor changes, as getting a good final audio result depends on changes to both (e.g. repositioning ambience sound sources, changing object attenuation values), but I can see value in separating the two options for people who want more granular control?

For example, maybe hypothetically syncing audio bus editors would be a performance hit, so separating would allow people to still sync scenes changes when they're not working on the audio.

**Comment by Calinou** [comment]
> For example, maybe hypothetically syncing audio bus editors would be a performance hit, so separating would allow people to still sync scenes changes when they're not working on the audio.

There's no performance hit in doing that. If you don't change the parameters, nothing will be sent through the debugger protocol.

**Comment by vitorbalbio** [comment]
This would be great. Currently need to change, test, change, test is very cumbersome.

**Comment by Faless** [comment]
@Calinou yeah, I think reusing `Sync Scene Changes` is the best course of action.

**Comment by EpsylonPT** [comment]
Any news if this ended up being implemented in a later version?

**Comment by chucklepie** [comment]
Sorry if this is the same thing as it sounds opposite to the requirement, but related :)

When I update the volume_db for buses in game I was expecting the buses in the Audio tab during a running game would update themselves to the current level, but they're just stuck on whatever level they are set during running. So when I adjust volume in game I cannot see which, if any, of my buses are being affected.

**Comment by ghostradiogames** [comment]
So I've also noticed this, and it's pretty unfortunate considering I would imagine nobody uses a mixer before playing their game and then hoping for the best.  But more than that, I cannot for the life of me get the meters to work at all - has that been an issue for anyone else? I'm testing on Godot 4 Beta 14 and there is no way to get the meters to light up, even though the audio **IS** playing.  I also can't mute or solo or do anything of value with the mixer during playback, though the audio does play.  I'm not even able to verify if the audio is playing on the correct bus.

**Comment by Calinou** [comment]
> But more than that, I cannot for the life of me get the meters to work at all - has that been an issue for anyone else? I'm testing on Godot 4 Beta 14 and there is no way to get the meters to light up, even though the audio IS playing.

Meters only work in the editor, not if the game is playing audio.

**Comment by ghostradiogames** [comment]
Ah yeah, I did manage to do some changes in code, but, it's just not ideal. I'm just adding my 2 cents that I'd like to see this added,  I guess.

**Comment by elvisish** [comment]
Just discovered this isn't possible and having to start/stop the game hundreds of times while fine-tuning reverb and delay. I think it would be possible to write a script to do this (maybe an extension) and I'm surprised that hasn't been done already. The script could just give you the current numbers and you could plug those back into the audio mixer, but it would be much more convenient if it didn't require an extension or example.

**Comment by elvisish** [comment]
Would it be possible to plan this for 3.6/4.3?

**Comment by AThousandShips** [comment]
Without a PR or someone presenting an implementation we can't plan for it, there's currently not any plans from the audio team to work on this directly, AFAIK

**Comment by elvisish** [comment]
> there's currently not any plans from the audio team to work on this directly

Any particular reason why not? Godot seems to pride itself on audio and music capabilities, @reduz just added a PR for music synchronisation last week, it's very unusual that audio bus volumes and effects are just about the only thing that _cant_ be changed and synched in real-time (even scripts can!)

**Comment by AThousandShips** [comment]
I can't answer that, you'll have to ask them 🙂 but it's much more likely to be implemented if someone picks it up, the whole team, not just the audio team, has a lot of work right now and I don't expect there will be a lot of room to pick up new tasks

**Comment by unlessgames** [comment]
This feature would be really nice to have!

Until someone implements it, here is a dirty helper script that will periodically check for file modifications on a bus layout file and apply it to the audio server if it has been changed. It's not as good as continuously updating parameters but still better than having to restart the game to hear any changes. 
Note, it won't work if you are modifying the bus layout at runtime.

> save as AudioBusReloader.gd and add to your main scene as a node

```gdscript
class_name AudioBusReloader extends Node

@export var bus_path := "res://default_bus_layout.tres"
@export_range(0.016, 1.0) var delay_seconds := 0.05

var last_modified = 0


func _ready() -> void:
	if OS.has_feature("editor"):
		_watch_bus()
	else:
		queue_free()


func _watch_bus() -> void:
	if not FileAccess.file_exists(bus_path):
		push_error("no file at '%s'" % bus_path)
		return

	var modified = FileAccess.get_modified_time(bus_path)
	if modified != last_modified:
		last_modified = modified
		var bus_layout = load(bus_path)
		if bus_layout is AudioBusLayout:
			AudioServer.set_bus_layout(bus_layout)
		else:
			push_error("file at '%s' is not a valid AudioBusLayout" % bus_path)
			return

	await get_tree().create_timer(delay_seconds).timeout
	_watch_bus()

```

**Comment by Bigfeta1** [comment]
Posting in 2025 for a similar request. Live VU metering during runtime would significantly reduce the time required to achieve a proper audio balance. 

Cheers

**Comment by AThousandShips** [comment]
@Bigfeta1 please create a dedicated proposal instead

**Comment by KoBeWi** [comment]
There is a linked pull request, but its solution is a bit overkill. I was investigating potential alternatives and came out with 2:

- Take advantage of live edit and forward AudioServer methods to the running game.

This one is very simple. Live edit works by forwarding redo operations and the debugger is already aware of AudioServer. Currently forwarding only works for Nodes and Resources, but it's easy to add support for AudioServer methods and the implementation would be below 50 lines. In fact, I already managed to get it working 😅

The problem is that it assumes that Audio tab is editing the current bus layout, which _might_ not be the case. Which brings the second solution...

- Make Audio editor edit the bus layout resource

Right now Audio editor directly modifies AudioServer and the AudioBusLayout resource is only involved when loading and saving. This can be changed so that the editor edits the resource instead, and the resource is synchronizing its changes with AudioServer. This would make the live editing work without changes, but requires duplicating some methods from AudioServer, which is more invasive change. However it solves the problem of multiple bus layouts.

___

Independently from the above, there is another problem - creating a Resource in the editor does not create it remotely. Which means that even if methods are synced properly, you won't be able to add bus effects. This can be fixed separately and will benefit not only the Audio editor.
